critwall/src/main/java/org/opencadc/critwall/Main.java

```diff
@@ -67,51 +67,198 @@
+import ca.nrc.cadc.util.StringUtil;
 import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
 import java.util.Map;
+import java.util.Set;
 import java.util.TreeMap;
 
+import javax.naming.NamingException;
 import org.apache.log4j.Level;
 import org.apache.log4j.Logger;
+import org.opencadc.inventory.db.SQLGenerator;
 import org.opencadc.inventory.storage.StorageAdapter;
 
 /**
  * Main entry point for critwall.
- * 
- * @author pdowler
  */
 public class Main {
     private static final Logger log = Logger.getLogger(Main.class);
+    
+    private static final String CONFIG_PREFIX = Main.class.getPackage().getName();
+    private static final String SQLGENERATOR_CONFIG_KEY = SQLGenerator.class.getName();
+    private static final String DB_SCHEMA_CONFIG_KEY = CONFIG_PREFIX + ".db.schema";
+    private static final String DB_CONFIG_KEY = CONFIG_PREFIX + ".db.url";
+    private static final String DB_USERNAME_CONFIG_KEY = CONFIG_PREFIX + ".db.username";
+    private static final String DB_PASSWORD_CONFIG_KEY = CONFIG_PREFIX + ".db.password";
+    private static final String BUCKETSEL_CONFIG_KEY = CONFIG_PREFIX + ".buckets";
+    private static final String NTHREADS_CONFIG_KEY = CONFIG_PREFIX + ".threads";
+    private static final String LOCATOR_SERVICE_CONFIG_KEY = CONFIG_PREFIX + ".locatorService";
+    private static final String LOGGING_CONFIG_KEY = CONFIG_PREFIX + ".logging";
 
     public static void main(String[] args) {
         try {
-            // get log level from config
-            Level cfg = Level.INFO;
             Log4jInit.setLevel("org.opencadc", Level.WARN);
             Log4jInit.setLevel("ca.nrc.cadc", Level.WARN);
-            
+
+            PropertiesReader pr = new PropertiesReader("critwall.properties");
+            MultiValuedProperties props = pr.getAllProperties();
+
+            if (log.isDebugEnabled()) {
+                log.debug("critwall.properties:");
+                Set<String> keys = props.keySet();
+                for (String key : keys) {
+                    log.debug("    " + key + " = " + props.getProperty(key));
+                }
+            }
+
+            // Set up logging before parsing file otherwise it's hard to report errors sanely
+            String logCfg = getSingleProperty(props, LOGGING_CONFIG_KEY);
+            Level cfg = Level.INFO;
+            if (StringUtil.hasLength(logCfg)) {
+                cfg = Level.toLevel(logCfg);
+            }
+
             Log4jInit.setLevel("org.opencadc.inventory", cfg);
             Log4jInit.setLevel("org.opencadc.tap", cfg);
             Log4jInit.setLevel("org.opencadc.reg", cfg);
-            
-            // parse config file and populate/assign these
+            Log4jInit.setLevel("org.opencadc.critwall", cfg);
+
+            log.debug("log level: " + cfg);
+            log.debug("properties: " + props.toString());
+
+            // parse config file
+            String errMsg = "";
```

Comment:
Use a StringBuilder and append when there are more than a couple of strings ops to do.

Reply:
ok